File: code/WRSN-CEC2021-MMC-MFEA/src/algo/time/TPopulation.java
package algo.time;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;

import problem.Configs;
import problem.ProblemManager;

public class TPopulation {

	private ArrayList<TIndividual> indivs;
	private TIndividual[] best;

	public TPopulation() {
		indivs = new ArrayList<TIndividual>();
		best = new TIndividual[ProblemManager.getTaskNumber()];
	}

	public void init() {
		indivs.clear();
		
//		for (int i = 0; i < ProblemManager.getTaskNumber(); i++) {
//			TIndividual indiv = new TIndividual();
//			int skill = i % ProblemManager.getTaskNumber();
//			indiv.setSkillFactor(skill);
//			indiv.initByGreedyTime();
//			indiv.setFitness(indiv.calculateFitness());
//			indivs.add(indiv);
//		}
		
		for (int i = indivs.size(); i < Configs.P_POP_SIZE; i++) {
			TIndividual indiv = new TIndividual();
			int skill = i % ProblemManager.getTaskNumber();
			indiv.setSkillFactor(skill);
			indiv.randomInit();
			indiv.setFitness(indiv.calculateFitness());
			indivs.add(indiv);
		}
	}

	public void selfUX(TIndividual par1, TIndividual par2) {
		int n = ProblemManager.USSD;
		for (int i = 0; i < n; i++) {
			if (Configs.rand.nextBoolean()) {
				double tmp = par1.getGene(i);
				par1.setGene(i, par2.getGene(i));
				par2.setGene(i, tmp);
			}
		}
	}

	/**
	 * simulated binary crossover
	 * 
	 * @param par1
	 * @param par2
	 * @return
	 */
	public ArrayList<TIndividual> crossover(TIndividual par1, TIndividual par2) {
		double u = Configs.rand.nextDouble();
		double beta;
		if (u <= 0.5) {
			beta = Math.pow(2 * u, 1.0 / (Configs.NC + 1));
		} else {
			beta = Math.pow(1.0 / (2 - 2 * u), 1.0 / (Configs.NC + 1));
		}

		TIndividual c1 = new TIndividual();
		TIndividual c2 = new TIndividual();

		for (int i = 0; i < ProblemManager.USSD; i++) {
			double g1 = 0.5 * ((1 + beta) * par1.getGene(i) + (1 - beta) * par2.getGene(i));
			double g2 = 0.5 * ((1 + beta) * par2.getGene(i) + (1 - beta) * par1.getGene(i));

			c1.setGene(i, Math.max(0, Math.min(1, g1)));
			c2.setGene(i, Math.max(0, Math.min(1, g2)));
		}

		ArrayList<TIndividual> offspring = new ArrayList<TIndividual>();
		offspring.add(c1);
		offspring.add(c2);
		return offspring;
	}

	public void updateEvaluation() {
		ArrayList<TIndividual> indivs = new ArrayList<TIndividual>();
		indivs.addAll(this.getIndivs());

		for (int k = 0; k < ProblemManager.getTaskNumber(); k++) {
			final int task = k;
			indivs.sort(new Comparator<TIndividual>() {
				@Override
				public int compare(TIndividual o1, TIndividual o2) {
					// TODO Auto-generated method stub
					return -Double.valueOf(o1.getFitness(task)).compareTo(o2.getFitness(task));
				}
			});

			for (int rank = 0; rank < indivs.size(); rank++) {
				TIndividual indiv = indivs.get(rank);
				indiv.setFactorialRank(k, rank + 1);
				double scalarFitness = 1.0 / indiv.getFactorialRank(k);
				if (k == 0 || indiv.getScalarFitness() < scalarFitness) {
					indiv.setScalarFitness(scalarFitness);
				}
			}

			if (best[k] == null || best[k].getFitness(k) < indivs.get(0).getFitness(k)) {
				best[k] = indivs.get(0);
			}
		}
	}

	public void executeSelection() {
		this.sortByScalarFitness();
		while (this.getIndivs().size() > Configs.P_POP_SIZE) {
			this.getIndivs().remove(this.getIndivs().size() - 1);
		}
	}

	public void sortByScalarFitness() {
		this.getIndivs().sort(new Comparator<TIndividual>() {
			@Override
			public int compare(TIndividual o1, TIndividual o2) {
				// TODO Auto-generated method stub
				return -Double.valueOf(o1.getScalarFitness()).compareTo(o2.getScalarFitness());
			}
		});
	}

	public ArrayList<TIndividual> getIndivs() {
		return indivs;
	}

	public void setIndivs(ArrayList<TIndividual> indivs) {
		this.indivs = indivs;
	}

	public TIndividual getIndividual(int index) {
		return indivs.get(index);
	}

	public TIndividual[] getBest() {
		return best;
	}

	public void setBest(TIndividual[] best) {
		this.best = best;
	}

	public TIndividual getBest(int task) {
		return best[task];
	}

	public void setBest(int task, TIndividual indiv) {
		best[task] = indiv;
	}

	public TIndividual geTIndividual(int index) {
		return indivs.get(index);
	}

	public void addIndividual(TIndividual indiv) {
		indivs.add(indiv);
	}

	public void addIndividuals(Collection<TIndividual> indivs) {
		this.indivs.addAll(indivs);
	}
}

File: algo/time/TPopulation.py
from problem.Configs import Configs
from algo.time.TIndividual import TIndividual
from problem.ProblemManager import ProblemManager

class TPopulation(object):
    """ generated source for class TPopulation """
    indivs: 'list[TIndividual]'
    indivs = []

    best: 'list[TIndividual]'
    best = []

    def __init__(self):
        """ generated source for method __init__ """
        pass

    def init(self):
        """ generated source for method init """
        self.indivs.clear()
        for i in range(len(self.indivs), Configs.P_POP_SIZE):
            indiv = TIndividual()
            skill = i % ProblemManager.getTaskNumber()
            indiv.skillFactor = skill
            indiv.randomInit()
            indiv.fitness = indiv.calculateFitness()
            self.indivs.append(indiv)

    def selfUX(self, par1: TIndividual, par2: TIndividual):
        """ generated source for method selfUX """
        n = ProblemManager.USSD
        for i in range(n):
            if Configs.rand.uniform(0,1) < 0.5:
                tmp = par1.gene[i]
                par1.gene[i] =  par2.gene[i]
                par2.gene[i] = tmp

    # 
    # 	 * simulated binary crossover
    # 	 * 
    # 	 * @param par1
    # 	 * @param par2
    # 	 * @return
    # 	 
    def crossover(self, par1: TIndividual, par2: TIndividual):
        """ generated source for method crossover """
        u = Configs.rand.uniform(0, 1)
        beta = 0.0
        if u <= 0.5:
            beta = pow(2 * u, 1.0 / (Configs.NC + 1))
        else:
            beta = pow(1.0 / (2 - 2 * u), 1.0 / (Configs.NC + 1))

        c1 = TIndividual()
        c2 = TIndividual()

        for i in range(ProblemManager.USSD):
            g1 = 0.5 * ((1 + beta) * par1.gene[i] + (1 - beta) * par2.gene[i])
            g2 = 0.5 * ((1 + beta) * par2.gene[i] + (1 - beta) * par1.gene[i])

            c1.gene[i] = max(0, min(1, g1))
            c2.gene[i] = max(0, min(1, g2))

        offspring: 'list[TIndividual]'
        offspring = []
        offspring.append(c1)
        offspring.append(c2)
        return offspring

    def updateEvaluation(self):
        indivs: 'list[TIndividual]'
        indivs = self.indivs.copy()

        for k in range(ProblemManager.getTaskNumber()):
            task = k
            indivs.sort(key=lambda indiv: -indiv.getFitness(task))
            for rank in range(len(indivs)):
                indiv = indivs[rank]
                indiv.factorialRank[k] =  rank + 1
                scalarFitness = 1.0 / indiv.factorialRank[k]
                if k == 0 or indiv.scalarFitness < scalarFitness:
                    indiv.scalarFitness = scalarFitness

            if self.best[k] == None or self.best[k].getFitness(k) < indivs[0].getFitness(k):
                self.best[k] = indivs[0]

    def executeSelection(self):
        self.sortByScalarFitness()
        while len(self.indivs) > Configs.P_POP_SIZE:
            self.indivs.pop(len(self.indivs)-1)

    def sortByScalarFitness(self):
        self.indivs.sort(key=lambda indiv: -indiv.scalarFitness)

    # def getIndivs(self):
    #     """ generated source for method getIndivs """
    #     return self.indivs

    # def setIndivs(self, indivs):
    #     """ generated source for method setIndivs """
    #     self.indivs = indivs

    # def getIndividual(self, index):
    #     """ generated source for method getIndividual """
    #     return self.indivs[index]

    # @overloaded
    # def getBest(self):
    #     """ generated source for method getBest """
    #     return self.best

    # @overloaded
    # def setBest(self, best):
    #     """ generated source for method setBest """
    #     self.best = best

    # @getBest.register(object, int)
    # def getBest_0(self, task):
    #     """ generated source for method getBest_0 """
    #     return self.best[task]

    # @setBest.register(object, int, TIndividual)
    # def setBest_0(self, task, indiv):
    #     """ generated source for method setBest_0 """
    #     self.best[task] = indiv

    # def geTIndividual(self, index):
    #     """ generated source for method geTIndividual """
    #     return self.indivs.get(index)

    # def addIndividual(self, indiv):
    #     """ generated source for method addIndividual """
    #     self.indivs.add(indiv)

    # def addIndividuals(self, indivs):
    #     """ generated source for method addIndividuals """
    #     self.indivs.addAll(indivs)

